Sources/MongoSwift/BSON/BSONValue.swift

```diff
@@ -779,6 +779,66 @@ public struct Timestamp: BSONValue, Equatable, Codable {
     }
 }
 
+enum BSONEqualsError: Error { case InvalidArrayArgument(String) }
+
+// See https://github.com/realm/SwiftLint/issues/461
+// swiftlint:disable cyclomatic_complexity
+// A helper function to test equality between two BSONValues. This function tests for exact BSON equality.
+// This means that, for example, differing types with equivalent value are not equivalent.
+// e.g.
+//  4.0 (Double) != 4 (Int)
+func bsonEquals(lhs: BSONValue, rhs: BSONValue) throws -> Bool {
+    validateBSONTypes(lhs: lhs, rhs: rhs)
+
+    switch (lhs, rhs) {
+    case (let l as Int, let r as Int): return l == r
+    case (let l as Int32, let r as Int32): return l == r
+    case (let l as Int64, let r as Int64): return l == r
+    case (let l as Double, let r as Double): return l == r
+    case (let l as Decimal128, let r as Decimal128): return l == r
+    case (let l as Bool, let r as Bool): return l == r
+    case (let l as String, let r as String): return l == r
+    case (let l as RegularExpression, let r as RegularExpression): return l == r
+    case (let l as Timestamp, let r as Timestamp): return l == r
+    case (let l as Date, let r as Date): return l == r
+    case (_ as MinKey, _ as MinKey): return true
+    case (_ as MaxKey, _ as MaxKey): return true
+    case (let l as ObjectId, let r as ObjectId): return l == r
+    case (let l as CodeWithScope, let r as CodeWithScope): return l == r
+    case (let l as Binary, let r as Binary): return l == r
+    case (let l as Document, let r as Document): return l == r
+    case (let l as [BSONValue?], let r as [BSONValue?]):
+        return try zip(l, r).reduce(true, {prev, next in try bsonEquals(lhs: next.0, rhs: next.1) && prev})
+    case (_ as [Any], _ as [Any]):
+        throw BSONEqualsError.InvalidArrayArgument(
+            "arrays not of type [BSONValue?] should not be compared with bsonEqual()"
+        )
+    default: return false
+    }
+}
+
+// A helper function to test equality between two BSONValue?s. See bsonEquals for BSONValues (non-optional) for more
+// information.
+func bsonEquals(lhs: BSONValue?, rhs: BSONValue?) throws -> Bool {
```

Comment:
This auxiliary method overloads the `bsonEquals()` implementation above. This seemed like a good idea for users that have optional `BSONValues` that they wished to compare, since those distinctly different types in Swift. This also makes the `[BSONValue?]` case much cleaner.

Reply:
Would we also need `bsonEquals()` to an optional `lhs` and non-optional `rhs`, and vice versa?

Reply:
In that case, would it make sense to do everything under a single `bsonEquals(lhs: BSONValue?, rhs: BSONValue?)` method?

Reply:
It should be fine to do it that way too. I just had them separated because it separated the logic for handling optional arguments away from the actual logic of handling the equality checking of `BSONValue`.

Do you think we should just merge these two functions?